### PR TITLE
a little reorg :-p

### PR DESCRIPTION
**Pull request by zocterminal**
Reorganized source code to consolidate copied functions and structures:
- common stuff is now in train_common.c and train_common.h
- cuda version needs to define TRAIN_CUDA before including the header or the .c file.
- using the checkFopen(), mallocCheck() etc. functions in train_gpt2.c also.

It may be possible to consolidate the reading of the model a bit more by using ifdefs with the TRAIN_CUDA and TRAIN_CPU macros.

Enjoy.

### CODE REVIEW

File: train_common.c [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+}
+
+
+// ----------------------------------------------------------------------------
+// Occasional gen_forward step
+
+void gen_forward(GPT2 *modelp, Tokenizer *tokenizerp, int genT, int B, int T  ) {
+
+    int* gen_tokens = (int*)mallocCheck(B * T * sizeof(int));
+#ifdef TRAIN_CUDA
```
Comment (karpathy) [inline train_common.c:227]: oh oh here we go. Instead of this maybe the `gen_forward` should be kept in the originals.

**Comment by karpathy** [comment]
So I want to do this, I'm just not sure when :)

**Comment by karpathy** [comment]
I certainly don't want to go down the IFDEF switching hell.

**Comment by rosslwheeler** [comment]
Yes, it would be great if we can reduce / eliminate the use of #ifdef's in the source.

File: train_common.c (hunk 1 above)
Comment (zocterminal) [reply train_common.c:227]: Putting the gen forward back would be easy (just delete the function and leave the original source in place).  And in that case the cudaChecks can go back as well.  I'm not at a build environment at the moment, but I can do that tomorrow.  Just tell me what you prefer to have changed. 

E.g. we could also eliminate the .h file and the changes in Makefile changes by just including the .c file in the other files (as you did in the test_gpt.cu, I just didn't know what style you prefer, so I did it my way (I'm used to compressing code and heavy use of ifdefs).

**Comment by zocterminal** [comment]
What's hell for some is just a cozy summer vacation to others ;-)

**Comment by dagelf** [comment]
Question is, really, what is the future of this project - how far can it go and how complex can it really get. Maybe split it out when it gets to the 4th target, 3 targets/flavors in one file doesn't mess with the elegant simplicity too much, does it? And exceptions for compilers wouldn't have any reason to blow up, would it? 

It is C after all...

**Comment by dagelf** [comment]
Having most of a file covered in ifdefs sure messes with the elegant simplicity. Compiler quirks should be few, so its the targets that are the real concern, isn't it? But it is C after all, maybe the pragmatic approach would be to refactor/split things out to separate files if and when a 4th platform is added? 2 seems managable, 3 might even be. Four could become too much to keep track of. I am here to learn though.

**Comment by zocterminal** [comment]
Closing (has hard to fix merge conflicts now, new version with different approach is https://github.com/karpathy/llm.c/pull/177).

**Pull request closed without merging**